Add early short-circuit evaluation for OR/AND in new analyzer to prevent unnecessary scalar subquery execution

fhw12345 opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Add early short-circuit evaluation for logical OR/AND expressions in the new analyzer. When resolving OR/AND functions, if any argument is a decisive constant (truthy for OR, falsy for AND), the entire expression is replaced with the constant value before argument resolution, preventing scalar subqueries in non-decisive branches from being executed. Resolves #83017

### Details
This PR adds early short-circuit evaluation for logical OR/AND expressions in `QueryAnalyzer::resolveFunction()`. When resolving OR/AND functions, the analyzer now checks if any argument is a decisive constant:

- For `OR`: if any argument is a truthy constant (1, true, "true"), replace the entire expression with 1
- For `AND`: if any argument is a falsy constant (0, false, "false"), replace the entire expression with 0

This optimization happens **before** argument resolution, preventing scalar subqueries in non-decisive branches from being executed at all.

Resolves #83017

haowenfeng123 committed:
Add early short-circuit evaluation for OR/AND

haowenfeng123 committed:
Add early short-circuit evaluation for OR/AND

haowenfeng123 committed:
fix style

haowenfeng123 committed:
fix style

haowenfeng123 committed:
fix style

fhw12345 commented:
Hi @alexey-milovidov could you help review this PR thanks!

haowenfeng123 committed:
add new setting

fhw12345 commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -2779,6 +2781,87 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+     * the remaining arguments – this prevents scalar
+     * sub-queries in those arguments from running.
+     */
+    if (function_name == "or" || function_name == "and")
```
@novikd It seems that `IFunctionBase::getConstantResultForNonConstArguments` can only be called after the `FunctionBase` has been built. However, at the top of `resolveFunction`, the `FunctionBase` isn't constructed yet. Is there another way to use this function at this point? I'd appreciate any suggestions, thanks!

fhw12345 replied:
@novikd Added a new setting `enable_function_early_short_circuit` to control whether to enable the short-circuit optimization and skip evaluation of scalar subqueries. Please take a look when you have a moment, thanks!

haowenfeng123 committed:
implement getConstantResultForNonConstArguments to do short-circuit

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

haowenfeng123 committed:
remove unused log

haowenfeng123 committed:
change comment

haowenfeng123 committed:
remove unused include

haowenfeng123 committed:
fix style

haowenfeng123 committed:
fix style

haowenfeng123 committed:
fix style

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Update SettingsChangesHistory.cpp

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Update Settings.cpp

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

novikd commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -2820,6 +2821,77 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
             node.getNullsAction() == NullsAction::IGNORE_NULLS ? "IGNORE" : "RESPECT");
 }
+
+/** Attempts to compute a constant result for any function by evaluating its arguments at compile time.
```
```suggestion
/** Attempts to compute a constant result for any function by evaluating its arguments at the query analysis phase.
```

novikd committed:
Update src/Core/Settings.cpp

novikd committed:
Update src/Analyzer/Resolve/QueryAnalyzer.cpp

novikd committed:
Update src/Analyzer/Resolve/QueryAnalyzer.cpp

novikd committed:
Update src/Analyzer/Resolve/QueryAnalyzer.cpp

novikd committed:
Update src/Analyzer/Resolve/QueryAnalyzer.cpp

novikd committed:
Update src/Core/Settings.cpp

fhw12345 committed:
Add new setting 'enable_function_early_short_circuit'

Added a new setting 'enable_function_early_short_circuit' to the settings changes history.

fhw12345 committed:
Remove empty line before settings changes

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

novikd reviewed (approved):
In general, LGTM. Let's see the result of CI

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Update SettingsChangesHistory.cpp

fhw12345 commented:
> In general, LGTM. Let's see the result of CI
@novikd Looks like the failure tests were not caused by my code change, could u help merge PR, thanks.

fhw12345 commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -2820,6 +2821,78 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+    if (resolver)
+    {
+        FunctionBasePtr base;
+        try
+        {
+            base = resolver->build(arg_columns);
+        }
+        catch (...)
+        {
+            return nullptr;
+        }
```
> @KochetovNicolai and I discussed it, and it does not seem reasonable to throw an exception on almost every function if this optimization is enabled

@novikd I think the try-catch block here is returning a nullptr instead of throwing an exception. I'm not entirely sure I understand your point—could you clarify what changes you'd like me to make to the logic?

fhw12345 replied:
@novikd gently ping for review again, thanks.

alexey-milovidov replied:
🕵 Resolved in the current implementation: the optimization no longer calls `resolver->build` with placeholder `UInt8` columns. It type-resolves a cloned expression through `resolveExpressionNode`; unsupported or value-dependent expressions fall back to normal analysis. The accompanying regression tests cover the preserved source expression and fallback cases.

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

haowenfeng123 committed:
add funtion check

haowenfeng123 committed:
Merge remote-tracking branch 'origin' into short-circuit_for_AND_OR

haowenfeng123 committed:
fix conflict

haowenfeng123 committed:
fix conflict

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Update SettingsChangesHistory.cpp

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Update SettingsChangesHistory.cpp

fhw12345 committed:
Merge branch 'master' into short-circuit_for_AND_OR

fhw12345 committed:
Add documentation for query plan optimizations

fhw12345 commented:
I’m extremely disappointed that this PR was half-reviewed and then ignored. Leaving it in limbo is blocking progress and needs immediate attention. @clickhouse-admin

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,97 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+                col.column = res->getColumn();
+                col.type = res->getResultType();
+            }
+            else col.type = std::make_shared<DataTypeUInt8>();
```
Fixed in 48990a316. The pre-resolution fold is now restricted to builtin `and`/`or`; type-dependent functions are no longer evaluated with `UInt8` placeholder types. Added regression coverage for `isNullable`, `isNull`, `toTypeName(DateTime64)`, and `defaultValueOfArgumentType(Nullable(...))`.

fhw12345 replied:
Fixed in 70082ded6. The early detector now returns only the decisive boolean value and no longer constructs a node from placeholder types. A cloned expression is resolved in type-only mode (scalar subqueries are analyzed but not executed), and the replacement constant is built with that clone's real Nullable/Bool result type. Added the two exact `toTypeName` regressions from this comment.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -559,6 +653,21 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+      */
+    if (scope.context->getSettingsRef()[Setting::enable_function_early_short_circuit])
+    {
+        auto const_res = getConstantResultFromFunctionArgs(node, scope);
```
Fixed in 48990a316. The optimization now runs after scoped-lambda lookup and is guarded against registered SQL and executable UDFs, so it cannot bypass a user-defined function with the same name. Added regression coverage for a scoped lambda shadowing `toTypeName`.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +674,28 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+        if (const_res)
+        {
+            auto value_string = const_res->getValueStringRepresentation();
+            node = std::move(const_res);
```
Fixed in 70082ded6. Before replacing the original subtree, a clone is resolved in type-only mode and checked with the same `hasAggregateFunctionNodes` / `hasFunctionNode(..., "arrayJoin")` safety conditions. The fold is rejected when either carrier is present. Added regressions for both `sum(number)` cardinality and `arrayJoin([1, 2])`.

fhw12345 commented on src/Analyzer/Resolve/evaluateScalarSubqueryIfNeeded.cpp:
```diff
@@ -104,7 +104,7 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, Iden
     /// WITH (SELECT path FROM table_with_paths) AS path SELECT * FROM s3(path, NOSIGN);
     /// or a default (empty) value substituted for a parameterized view parameter.
-    const bool only_analyze_subquery = only_analyze
+    const bool only_analyze_subquery = (only_analyze || early_short_circuit_type_inference_in_process)
```
Fixed in 9199ffd32. `execute_for_exists` now always takes the literal conversion path, including during early-short-circuit type-only analysis, so its caller can safely require a `ConstantNode`. Added Array and Tuple EXISTS regressions (`0 AND exists(SELECT [1])` and `1 OR exists(SELECT tuple(1))`). I also audited the call sites; this is the only call passing `execute_for_exists = true`.

fhw12345 commented on tests/queries/0_stateless/03562_short_circuit_for_and_or.sql:
```diff
@@ -0,0 +1,69 @@
+SELECT defaultValueOfArgumentType(CAST(NULL AS Nullable(UInt8))) IS NULL;
+
+SELECT 'Test scoped lambda shadows a builtin function';
+WITH (x -> 42) AS toTypeName SELECT toTypeName(number) FROM numbers(1);
```
Fixed in 7c31f7409. The regression now shadows the optimized builtin `or` itself and passes a decisive constant: `WITH ((x, y) -> 42) AS `or` SELECT `or`(1, number)`. Without the scoped-lambda guard the early fold would incorrectly return `1`; the expected lambda result is `42`.

fhw12345 commented on tests/queries/0_stateless/03562_short_circuit_for_and_or.sql:
```diff
@@ -0,0 +1,69 @@
+SELECT 'Check the read_rows of the above queries to ensure that the short circuit is working';
+SYSTEM FLUSH LOGS query_log;
+
+SELECT read_rows FROM system.query_log WHERE current_database = currentDatabase() AND query LIKE '%SELECT 1 where 0 OR ((SELECT count(*) FROM test_03562) > 1 AND 1) AS bool%' AND type = 'QueryFinish' AND is_initial_query = 1 ORDER BY event_time DESC LIMIT 1;
```
Fixed in 7c31f7409. Removed the stale `query_log` probe. The test now has 11 executed-query probes and exactly 11 corresponding reference rows.

fhw12345 commented on src/Analyzer/Resolve/evaluateScalarSubqueryIfNeeded.cpp:
```diff
@@ -104,7 +104,7 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, Iden
     /// WITH (SELECT path FROM table_with_paths) AS path SELECT * FROM s3(path, NOSIGN);
     /// or a default (empty) value substituted for a parameterized view parameter.
-    const bool only_analyze_subquery = only_analyze
+    const bool only_analyze_subquery = (only_analyze || early_short_circuit_type_inference_in_process)
```
Fixed in 994e356d0 (with the style follow-up in 824f31afd). Type-only scalar subqueries now become non-constant placeholders carrying only the real type, never sample-block default values. If a value-sensitive function such as `tupleElement` requires the actual constant, clone resolution fails safely and the optimization falls back to normal scalar evaluation. Added a focused regression with `tupleElement` and a scalar index.

fhw12345 commented on src/Analyzer/Resolve/evaluateScalarSubqueryIfNeeded.cpp:
```diff
@@ -104,7 +105,7 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, Iden
     /// WITH (SELECT path FROM table_with_paths) AS path SELECT * FROM s3(path, NOSIGN);
     /// or a default (empty) value substituted for a parameterized view parameter.
-    const bool only_analyze_subquery = only_analyze
+    const bool only_analyze_subquery = (only_analyze || early_short_circuit_type_inference_in_process)
```
Fixed in 0a12c597b. Early type inference now proves scalar cardinality before allowing a fold: aggregation without GROUP BY and constant LIMIT <= 1 are accepted; unproven Query/Union cardinality marks inference as failed and forces normal execution. Added the exact `numbers(2)` regression expecting `INCORRECT_RESULT_OF_SCALAR_SUBQUERY`.

fhw12345 commented on src/Analyzer/Resolve/evaluateScalarSubqueryIfNeeded.cpp:
```diff
@@ -368,15 +371,30 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, Iden
-        || !context->hasQueryContext() || !nearest_query_scope || parameterized_view_arguments_in_resolve_process)
+    /// The EXISTS caller also requires a ConstantNode so it can derive the boolean result from
+    /// whether the scalar subquery returned NULL; this must hold in type-only analysis as well.
+    if (execute_for_exists || !context->getSettingsRef()[Setting::enable_scalar_subquery_optimization]
```
Fixed in 0a12c597b. `execute_for_exists` now always marks early type inference as failed, so its type-only sample value can never authorize a fold. The clone may still use a temporary ConstantNode to satisfy the existing caller contract, but the original expression is resolved normally. Added the empty-EXISTS/tupleElement regression expecting `ARGUMENT_OUT_OF_BOUND`.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +639,72 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+                    early_short_circuit_type_inference_failed = previous_type_inference_failure;
+                });
+
+                resolveExpressionNode(
```
Fixed in 45b5f09c6. Speculative resolution now uses an isolated copy of `IdentifierResolveScope`: the parent link is removed, identifier lookup state/cache is cleared and permanently disabled, and the projection-mask map is copied. In-progress stack entries and placeholder alias resolutions are therefore discarded with the temporary scope on both success and exception paths. Added a WITH scalar-alias + `tupleElement` fallback regression (using `assumeNotNull` so the normal query is valid).

fhw12345 replied:
Follow-up in 6ae288ce4 strengthens the isolation: speculative resolution now runs in a separate temporary `QueryAnalyzer`, so all analyzer-global caches/maps/counters are discarded as well. It also skips speculation entirely for expressions with outer-scope Identifier/Column dependencies.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +655,83 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+      * user-defined function with the same name.
+      */
+    if (!early_short_circuit_type_inference_in_process
+        && scope.context->getSettingsRef()[Setting::enable_function_early_short_circuit]
```
Fixed in 614470131. Analyzer-time folding is now gated on `short_circuit_function_evaluation != DISABLE`. Added a regression where `1 OR throwIf(1)` must still throw when short-circuit evaluation is explicitly disabled.

fhw12345 replied:
Test follow-up in bf7e853e0: changed the regression to `1 OR (SELECT throwIf(1))`. Plain `throwIf(1)` is already eliminated by existing constant folding independently of the runtime setting; the scalar-subquery form specifically verifies that this new analyzer-time optimization is disabled and therefore executes the scalar.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,95 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+
+bool isSafeCountScalarSubqueryForEarlyShortCircuit(const QueryNode & query)
+{
+    if (query.hasGroupBy()
```
Fixed in e17794fa2. The safe-count preflight now recursively rejects any nested Query/Union anywhere under the scalar query (including table-function and parameterized-view arguments), before constructing the temporary analyzer. Added normal-result and `throwIf` regressions for a scalar nested inside `numbers(...)`.

fhw12345 replied:
Fixed in 6e34b0139. The isolated type-inference scope now clears expression arguments, local/global aliases, and CTE mappings before resolving. Therefore hidden scalar work cannot be expanded from a parent alias/CTE in the temporary analyzer; unresolved dependencies force the regular path. Added successful-value and `throwIf` alias regressions for `numbers(assumeNotNull(idx))`.

fhw12345 replied:
Additional hardening in 331c60bc5: `early_short_circuit_type_inference_in_process` now makes `only_analyze_subquery` unconditionally true, overriding the table-function/parameterized-view execution exceptions. Speculative analysis can therefore never execute a scalar; contexts requiring a real value fail and fall back.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,114 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+        || function_node->isWindowFunction())
+        return {};
+
+    const bool decisive_value = function_name_to_fold == "or";
+    for (const auto & argument : function_node->getArguments().getNodes())
+    {
+        std::optional<bool> argument_value;
+        if (const auto * constant_node = argument->as<ConstantNode>())
+        {
+            const auto & type = constant_node->getResultType();
+            const auto value = constant_node->getValue();
+            if (isNativeNumber(removeNullable(type)) && !value.isNull())
+                argument_value = applyVisitor(FieldVisitorConvertToNumber<bool>(), value);
+        }
+        else if (argument->as<FunctionNode>())
+        {
+            argument_value = getEarlyShortCircuitResultForAndOr(argument, function_name_to_fold);
+        }
+
+        if (argument_value && *argument_value == decisive_value)
+            return decisive_value;
```
Fixed in 269d79bab. Decisive-constant detection is now strictly left-to-right: the first unresolved argument stops the optimization, so it can only prune a suffix after a decisive prefix. Added `throwIf` regressions for live scalar prefixes before trailing `OR 1` / `AND 0`.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +707,85 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+
+                auto value_string = const_res->getValueStringRepresentation();
+                node = std::move(const_res);
+                return { value_string };
```
Fixed in 269d79bab. The fast path now mirrors regular folding by constructing `ConstantNode(ConstantValue, source_expression, deterministic)`, returning the projection names from the isolated normal-resolution pass, and synchronizing the subquery counter so headers keep the same `_subquery_N` numbering.

alexey-milovidov replied:
🕵 Addressed in c6a6a9596c27. Type-only scalar analysis now preserves the normal `_subquery_N` projection name, and the focused `TSVWithNames` test verifies that no internal placeholder reaches the header.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +712,88 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+
+            if (type_inference_succeeded
+                && !hasAggregateFunctionNodes(node_for_type_inference)
+                && !hasFunctionNode(node_for_type_inference, "arrayJoin"))
```
Fixed in 8aa957d85. Added a subquery-boundary-aware late-carrier detector for `grouping`, window functions, and WindowNode. It gates both the early fast path and the existing regular constant-folding path, so neither can erase constructs validated by the enclosing query. Added `grouping` error and window-row regressions.

fhw12345 replied:
Correction after CI in a780b442f: ClickHouse intentionally allows existing constant folding to remove constant window/grouping branches; `03905_window_function_constant_in_group_by` explicitly requires this behavior and failed when the proposed late-carrier guard was added to the regular path. The guard and its incorrect regressions were reverted so the fast path remains aligned with established constant-folding semantics. Aggregate and `arrayJoin` guards remain unchanged.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,119 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+        }
+        else if (argument->as<FunctionNode>())
+        {
+            argument_value = getEarlyShortCircuitResultForAndOr(argument, function_name_to_fold);
```
Fixed in 2aebbca5d. Early decisive-prefix detection no longer recurses into nested FunctionNodes at all. Nested `and`/`or` calls are resolved independently through their own `resolveFunction`, where lambda/SQL/executable UDF shadowing is known. Added the exact nested scoped-`or` wrong-result regression (expected 0).

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +708,88 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+            type_inference_scope.projection_mask_map
+                = std::make_shared<std::map<IQueryTreeNode::Hash, size_t>>(*scope.projection_mask_map);
+
+            QueryAnalyzer type_inference_analyzer(/*only_analyze_=*/ false);
```
Fixed in d39e1debe. Before creating the temporary analyzer, the entire candidate subtree is scanned through FunctionFactory; any function that is not deterministic globally or within the query disables the optimization. This avoids speculative random evaluation rather than trying to synchronize `functions_cache`. Added a safe-count/WHERE regression with `randConstant()` whose normal path must execute `throwIf`.

fhw12345 replied:
Fixed in 11f3a5011. The preflight now treats every function name not registered in builtin `FunctionFactory` as unsafe, except for the aggregate `count` already validated by the strict safe-scalar matcher. SQL, executable, WebAssembly, and other hidden UDF bodies therefore never enter speculative analysis. Builtins must also be deterministic globally and within the query.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +741,90 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+            type_inference_scope.identifier_in_lookup_process.clear();
+            type_inference_scope.clearIdentifierCache();
+            type_inference_scope.disableIdentifierCachePermanently();
+            type_inference_scope.expression_argument_name_to_node.clear();
+            type_inference_scope.aliases = {};
+            type_inference_scope.global_with_aliases = {};
+            type_inference_scope.cte_name_to_query_node.clear();
```
Fixed in cd5fb6c3f. Preflight now checks every FunctionNode name against local/global function aliases across the scope chain before builtin lookup. Any scoped lambda binding makes the candidate unsafe. Added an `abs`-shadowing regression with a String-returning lambda, which must preserve the normal ILLEGAL_TYPE error.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +741,90 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+                auto result_type = node_for_type_inference->getResultType();
+                auto result_column = result_type->createColumnConst(1, static_cast<UInt8>(*short_circuit_result));
+                ConstantValue constant_value{ std::move(result_column), std::move(result_type) };
+                node = std::make_shared<ConstantNode>(std::move(constant_value), node, true /*is_deterministic*/);
```
Fixed in cd5fb6c3f. The folded ConstantNode now uses the fully resolved/masked isolated clone as `source_expression`, never the untouched live tree. Secret-argument discovery therefore runs before the source is retained, while projection metadata remains from the same isolated normal-resolution pass.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,166 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+    const auto & join_tree = query.getJoinTreeNode();
+    if (!join_tree
+        || (join_tree->getNodeType() != QueryTreeNodeType::TABLE
+            && join_tree->getNodeType() != QueryTreeNodeType::IDENTIFIER)
```
Fixed in 98bebb2bc. Plain table identifiers are now checked against CTE and table-alias maps across the entire parent scope chain before speculation. A visible CTE/table alias makes the count scalar unsafe even if a same-named catalog table exists. Added the collision regression with a real table plus invalid CTE.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +759,91 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+                auto result_column = result_type->createColumnConst(1, static_cast<UInt8>(*short_circuit_result));
+                ConstantValue constant_value{ std::move(result_column), std::move(result_type) };
+                node = std::make_shared<ConstantNode>(
+                    std::move(constant_value), std::move(node_for_type_inference), true /*is_deterministic*/);
```
Fixed in 98bebb2bc. The source expression is now a separate clone of the original subtree, so it retains Query/Union nodes and never contains `__early_short_circuit_scalar`. Secret mask decisions are collected from the fully resolved/masked analysis clone (including ConstantNode source chains) and applied by hash to the original-form source clone before it is stored.

fhw12345 commented on tests/queries/0_stateless/03562_short_circuit_for_and_or.sql:
```diff
@@ -0,0 +1,104 @@
+SELECT true AND (true OR (SELECT count(*) FROM test_03562) > 1) AS bool; -- true
+SELECT false OR (false AND (SELECT count(*) FROM test_03562) > 1) AS bool; -- false
+
+SELECT 'Test type-dependent functions use resolved argument types';
```
Fixed in 31aa1f695 / d38d0b2e8. Added an executable no-fasttest shell regression that asserts EXPLAIN SYNTAX retains the original scalar table name and never emits `__early_short_circuit_scalar`, and EXPLAIN QUERY TREE contains `[HIDDEN` while never containing the raw AES key when secret display is disabled.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +103,226 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+    std::map<IQueryTreeNode::Hash, size_t> & projection_mask_map)
+{
+    if (auto * constant = node->as<ConstantNode>();
+        constant && masked_hashes.contains(constant->getTreeHash()))
```
Fixed in 15c6970ac / 2865c6fd5. Mask propagation now aligns the resolved and original clones by node position, unwrapping folded ConstantNode source expressions as needed. Hashes are used only to allocate an id after the exact source constant is selected. The no-fasttest regression now uses the same literal once visibly and once as the AES key, asserting exactly one visible occurrence plus `[HIDDEN]`.

alexey-milovidov commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +818,94 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+        auto short_circuit_result = getEarlyShortCircuitResultForAndOr(node, function_name);
+        if (short_circuit_result
+            && !hasScopeDependentNodesForEarlyShortCircuit(node, scope)
+            && !hasUnsafeFunctionForEarlyShortCircuit(node, scope.context, scope))
```
🕵 Resolved the function tree in the speculative analysis and reject the early fold when any non-root ordinary function opts out of lazy short-circuit execution. Added regression coverage for `ignore(sleep(1))`.

fhw12345 replied:
Addressed by Alexey in bb2f3daee and refined in f7ba97771/e807871f1. Non-lazy functions such as `ignore` remain rejected. The only exception is a six-name comparison whitelist when the resolved comparison actually contains the internal placeholder for an already-approved single-row count scalar; this restores the PR's target read_rows optimization without broadening lazy eligibility.

fhw12345 replied:
Fixed in 19a50b68e. The comparison exception now requires exactly one operand to contain the approved scalar placeholder and the other operand to be a deterministic literal ConstantNode with no source expression. `throwIf(1)` and any computed/non-literal opposite side therefore keep the comparison non-lazy and force normal execution. Added the exact `count(*) > throwIf(1)` regression.

alexey-milovidov commented:
🕵 The enabled-path regression test already provides a focused measurement for this optimization: `03562_short_circuit_for_and_or` checks `system.query_log.read_rows` after the dead scalar `count()` cases. The folded `OR`/`AND` cases read `0` rows, whereas the equivalent live or fallback cases read `1` row. This demonstrates that the opt-in analyzer path avoids executing the scalar subquery rather than merely changing its result.

alexey-milovidov commented on tests/queries/0_stateless/04700_early_short_circuit_source_expression.sh:
```diff
@@ -0,0 +1,69 @@
+. "$CUR_DIR"/../shell_config.sh
+
+$CLICKHOUSE_CLIENT -q "DROP TABLE IF EXISTS t_04700_early_short_circuit_source"
+$CLICKHOUSE_CLIENT -q "CREATE TABLE t_04700_early_short_circuit_source (x UInt8) ENGINE = Memory"
+
+syntax_output=$($CLICKHOUSE_CLIENT -q "
+    EXPLAIN SYNTAX
+    SELECT 1 OR ((SELECT count(*) FROM t_04700_early_short_circuit_source) > 0)
+    SETTINGS enable_function_early_short_circuit = 1
+")
+
+if grep -qF 't_04700_early_short_circuit_source' <<< "$syntax_output" \
+    && ! grep -qF '__early_short_circuit_scalar' <<< "$syntax_output"; then
+    echo "syntax_source_ok"
+else
+    echo "syntax_source_failed"
+fi
+
+header_output=$($CLICKHOUSE_CLIENT -q "
+    SELECT 1 OR ((SELECT count(*) FROM t_04700_early_short_circuit_source) > 0)
+    SETTINGS enable_function_early_short_circuit = 1
+    FORMAT TSVWithNames
+")
```
🕵 Fixed in 243775cbd49e: both the `EXPLAIN SYNTAX` and `TSVWithNames` probes now set `enable_analyzer = 1`, so they exercise the analyzer path.

alexey-milovidov commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,285 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+        const auto * resolved_constant = resolved_node->as<ConstantNode>();
+        if (!resolved_constant || !resolved_constant->hasSourceExpression())
+            return;
+        resolved_node = resolved_constant->getSourceExpression();
```
🕵 Addressed in 8acd04adbb05: when a masked folded `ConstantNode` is unwrapped to align with a non-constant source subtree, every constant in that subtree receives a secret mask. The regression now covers a folded `concat('SE', 'CRET')` encryption key.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,309 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+{
+    const auto & join_tree = query.getJoinTreeNode();
+    if (!join_tree
+        || (join_tree->getNodeType() != QueryTreeNodeType::TABLE
```
Fixed in 19a50b68e. Before approving a TABLE/IDENTIFIER source, the preflight resolves the catalog table and rejects any storage whose `isView()` is true (also failing closed when the source cannot be resolved). Added a view-backed count regression with `throwIf(1)` in the view query.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,401 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+    if (first_is_scalar == second_is_scalar)
+        return false;
+
+    const auto & other_argument = arguments[first_is_scalar ? 1 : 0];
```
Fixed in 65d855899. The comparison exemption now accepts only a direct `_subquery_N` ColumnNode operand; recursive placeholder detection was removed. Any wrapper on the scalar side (`concat`, `toString`, `sleep`, casts, etc.) leaves the comparison non-lazy and forces normal execution. The opposite operand must still be a plain deterministic literal.

fhw12345 commented on tests/queries/0_stateless/03562_short_circuit_for_and_or.sql:
```diff
@@ -0,0 +1,115 @@
+SELECT 1 OR ((SELECT count() FROM numbers(1) WHERE throwIf(randConstant() % 1 = 0) = 0) > 0); -- { serverError FUNCTION_THROW_IF_VALUE_IS_NON_ZERO }
+
+SELECT 'Test comparison non-placeholder expressions stay eager';
+SELECT 1 OR ((SELECT count(*) FROM test_03562) > throwIf(1));
```
Fixed in 17c7a58e/75fe3bd3. The exact non-literal comparison query now has a unique alias and a query_log assertion requiring `read_rows = 10001`, so the regression distinguishes early pruning from the regular analyzer path instead of relying only on the result value.

fhw12345 commented on tests/queries/0_stateless/04700_early_short_circuit_source_expression.sh:
```diff
@@ -0,0 +1,71 @@
+
+secret_output=$($CLICKHOUSE_CLIENT -q "
+    EXPLAIN QUERY TREE
+    SELECT 1 OR notEmpty(encrypt('aes-128-ecb', 'x', concat('KEYPART1', 'KEYPART2')))
```
Fixed in de3678173. The masking probe now uses the same valid 16-byte literal once as visible concat input and once as the AES key under EXPLAIN SYNTAX. It requires `[HIDDEN]` and exactly one visible literal occurrence, covering the prior hash-vs-position regression.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +983,95 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+            if (type_inference_succeeded
+                && !hasAggregateFunctionNodes(node_for_type_inference)
+                && !hasFunctionNode(node_for_type_inference, "arrayJoin")
+                && !hasFunctionNotSuitableForEarlyShortCircuit(node_for_type_inference))
```
Fixed in f872d409/3804afb9/2fd6b8b4. The lazy-eligibility walker now exempts nested builtin `and`/`or` nodes themselves while continuing to recurse into their descendants, so unsafe inner functions are still rejected. Added `1 OR (0 OR count_scalar > 0)` with a query_log assertion requiring `read_rows = 1`.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +985,95 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+            if (type_inference_succeeded
+                && !hasAggregateFunctionNodes(node_for_type_inference)
+                && !hasFunctionNode(node_for_type_inference, "arrayJoin")
+                && !hasFunctionNotSuitableForEarlyShortCircuit(node_for_type_inference))
```
Fixed comprehensively in b97ab7078/d05f256f6. Added a tree-wide invariant: an early scalar placeholder is legal only as a direct operand of one of the six comparison functions, with the opposite operand a deterministic source-less literal. Any placeholder under intDiv/modulo/cast/concat/isNull/other wrapper, or used directly by logical functions, rejects speculative folding. Added the exact empty-count `intDiv(1, 0)` regression expecting ILLEGAL_DIVISION.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,461 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+            && join_tree->getNodeType() != QueryTreeNodeType::IDENTIFIER)
+        || (join_tree->getNodeType() == QueryTreeNodeType::IDENTIFIER
+            && isTableIdentifierShadowedInScope(join_tree->as<IdentifierNode &>(), scope))
+        || isViewBackedSource(join_tree, scope)
```
Fixed in 711428c2b. The safe-count preflight now fails closed when `additional_table_filters` is configured or the source table has a non-trivial SELECT row policy, so late planner-attached filters cannot be skipped. Added a regression whose additional filter invokes `throwIf(1)`, verifying the regular path still raises the error.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +101,452 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+        {
+            /// The safe scalar-subquery check separately verifies the aggregate count call.
+            /// Any other unknown name may be a SQL/executable UDF whose body is not visible here.
+            if (function->getFunctionName() != "count")
```
Fixed in bb31f9279. The count exception is now scoped to a QueryNode that passed the strict safe scalar-count preflight, so a normal aggregate count() remains on the regular path. Added regressions for both 1 OR count() and 0 AND count() over numbers(10).

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -596,6 +1054,96 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
+                copySecretMasksByPosition(node_for_type_inference, source_expression, *scope.projection_mask_map);
+
+                ConstantValue constant_value{ std::move(result_column), std::move(result_type) };
+                node = std::make_shared<ConstantNode>(
```
Fixed in 120d1a663. Early AND/OR folding is disabled while resolving JOIN ON expressions, preventing a raw scalar-subquery source expression from being unwrapped into PlannerActionsVisitor. Added the reported JOIN ON scalar-subquery regression.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +104,490 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+        || (join_tree->getNodeType() == QueryTreeNodeType::IDENTIFIER
+            && isTableIdentifierShadowedInScope(join_tree->as<IdentifierNode &>(), scope))
+        || isViewBackedSource(join_tree, scope)
+        || hasLateAttachedTableFilter(join_tree, scope)
```
Fixed in fb516da63. The safe count-scalar preflight now rejects view-backed and remote storages. Remote tables can attach shard-local policies or filters that initiator metadata cannot prove safe, so the normal scalar execution path is preserved.

alexey-milovidov commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +104,492 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+{
+    /// Additional filters are parsed only by the planner, after the speculative type-only
+    /// analysis. Any configured filter may affect the selected table, so fail closed.
+    if (!scope.context->getSettingsRef()[Setting::additional_table_filters].value.empty())
```
🕵 Addressed in f5061641deec: the preflight now rejects `additional_table_filters` from the scalar subquery context as well as the outer scope, and `03562_short_circuit_for_and_or` covers the subquery-local `SETTINGS` case.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +104,497 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+
+    /// Remote storages can attach policies and filters on shard-local tables that are not
+    /// visible in initiator-side metadata, so they cannot be proven safe speculatively.
+    return table->getStorage()->isView() || table->getStorage()->isRemote();
```
Fixed in 5bb68025f. The scalar-count source preflight now explicitly rejects StorageAlias, which forwards reads and policy evaluation to an independently changing target without forwarding isView(). Added a regression for an Alias pointing at a throwIf-backed View; the dead branch now preserves the normal exception.

fhw12345 commented on src/Analyzer/Resolve/resolveFunction.cpp:
```diff
@@ -98,6 +104,509 @@ void checkFunctionNodeHasEmptyNullsAction(FunctionNode const & node)
+    /// Alias forwards reads to another storage but does not forward isView(), and policies are
+    /// evaluated against its target instead of the alias name. Reject it rather than proving a
+    /// target that can change independently during analysis.
+    return storage->isView() || storage->isRemote() || storage->getName() == "Alias";
```
Fixed in 234bde95c. The opt-in scalar-count fold is now limited to local physical Memory and MergeTree-family tables. Merge, Alias, views, remote storages, and every other storage without a dedicated safety proof fail closed. Added the reported Merge-over-throwIf-view regression.
